### ISSUE TITLE
Rule proposal: prefer `as const`

### ISSUE THREAD

**Issue by mohsen1** [typescript-eslint/typescript-eslint#312]
In a lot of our code we have to cast string literals to literals types to conform to types so assignments work:

```ts
  type Styles = {
    position: 'relative' | 'absolute'
  }
  
  let styles = {
    position: 'relative' as 'relative'
  }
  
  let s: Styles = styles 
```

This is more common when using React HOCs but above example demonstrate the problem well enough. 

[With TypeScript 3.4 we'll have the `as const` literal expression](https://github.com/Microsoft/TypeScript/pull/29510) which should be preferred to string or number literal types in this case because changing the value does not require changing the type.

Our example code would be better written as:

```ts
type Styles = {
  position: "relative" | "absolute";
};

let styles = {
  position: "relative" as const
};

let s: Styles = styles;
```

**More examples**:

```ts
let a: 'str' = 'str'; // bad 
let a = <'str'>'str'; // bad 
let a = 'str' as 'str' //bad 

let a = 'str' as const // good
```

```ts
// bad
let o = {
    foo: 'string' as 'string'
}

// good
let o = {
    foo: 'string' as const
}
```

### Rule name
Please suggest rule name. I'm not sure how this rule should be named

### Options
I can't think of any configurability for this rule

**Comment by bradzacher** [comment]
on question i'd have, is it worth hoisting the const up if we detect multiple within the same object?

i.e.
```TS
const styles = {
	display: 'flex' as 'flex',
    position: 'absolute' as 'absolute',
};

// does it fix to
const styles = {
	display: 'flex' as const,
    position: 'absolute' as const,
};
// pr
const styles = {
	display: 'flex',
    position: 'absolute',
} as const;
```

I have only briefly toyed with `as const` via the next tag, so I don't know if the latter example will do what I think it does..

**Comment by ckknight** [comment]
I don't think it'd be appropriate to fix to moving the `as const` to the object, as that has a dramatically different meaning (that all the properties are `readonly`)
